### PR TITLE
[HandshakeToFIRRTL] Add support for PackOp and UnpackOp

### PR DESCRIPTION
**Pull request by Dinistro**
This PR introduces lowerings of the operations working on tuples, related to #3037.

`TupleTypes` are transformed to `BundleTypes` with element names `field0`, ..., `field[N-1]`.
The pack and unpack operations connect the separated values with the ones encapsulated in the bundle while having similar control logic as `ForkOp` and `JoinOp`.

### CODE REVIEW

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp
Comment (mortbopet) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:268]: While it wasn't there before, a comment to preface this function would be a nice addition.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp
Comment (mortbopet) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:280]: as above.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp
Comment (mortbopet) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:339]: `else if (auto tupleType = type.dyn_cast<TupleType>()) { ...`

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp
Comment (mortbopet) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:353]: as above

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp
Comment (mortbopet) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:342]: ~~Should this be interleaved with `_`?~~ i see that `getTypeName` returns a `_` prefixed name... bad! But let's leave that up to a follow-up commit.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp
Comment (mortbopet) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:2804]: no brackets

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp
Comment (mortbopet) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:2814]: no brackets

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp
Comment (mortbopet) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:2831]: no brackets

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp
Comment (mortbopet) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:2841]: no brackets; `auto&& [element, output]` probably works here.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp
Comment (mortbopet) [inline lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:2844]: Side note - ughhh, can't wait until the `handshake-to-hw` rewrite gets up to speed, so assumptions like this can be reworked.

**Review by mortbopet — CHANGES_REQUESTED** [review]
Nice! Have you also lowered it all the way to hardware and tested it in a simulator? I've over time found a few bugs in the FIRRTL lowerings this way since we (unfortunately) don't have a nice way of testing actual hardware behavior (yet) for handshake operations in the CIRCT tests.

File: lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp
Comment (Dinistro) [reply lib/Conversion/HandshakeToFIRRTL/HandshakeToFIRRTL.cpp:2844]: +1
This is especially weird because it depends on a trait/interface added somewhere else. If someone removes `HasClock` everything explodes without any reasonable warnings :S

**Comment by Dinistro** [comment]
Thanks for the comments. I somehow always forget to drop the brackets of single statement loops...

We are currently setting up a simulation environment and we plan to test all the parts related to my project. If you prefer, we can delay merging this as soon as we have confirmation that simulations work as expected.

**Comment by mortbopet** [comment]
> We are currently setting up a simulation environment and we plan to test all the parts related to my project. If you prefer, we can delay merging this as soon as we have confirmation that simulations work as expected.

Sounds good - since you are (most likely) the only current user of these ops, then there's no rush to merge.

**Review by mortbopet — APPROVED** [review]

**Comment by mikeurbach** [comment]
> since we (unfortunately) don't have a nice way of testing actual hardware behavior (yet) for handshake operations in the CIRCT tests.

We do have some integration tests with verilator: https://github.com/llvm/circt/blob/main/integration_test/Dialect/Handshake/dot.mlir.

**Comment by Dinistro** [comment]
I can look into that, as I either way have to use simulation for other things.

**Comment by Dinistro** [comment]
There are some complications with buffers and memories. Currently, the lowering pass assumes that the values to store aren't bundle types, which is no longer true. 

Should I add support for buffers/memories in this PR or should this be split into a second PR?

**Comment by Dinistro** [comment]
I split up the PR into two parts. This PR now depends on #3108

**Comment by Dinistro** [comment]
This should now be ready.

**Review by mortbopet — APPROVED** [review]
LGTM! Thank you for adding the integration test :+1:.

**Comment by Dinistro** [comment]
You're welcome. Thanks for the very constructive reviews that help improve my development workflow :smile:.

I suggest that we wait for @mikeurbach as there is no rush to merge this.

**Pull request merged**
